_posts/2019-12-06-integration-testing-middleware.adoc

```diff
@@ -0,0 +1,695 @@
+<3> the system under test (SUT) invocation is extracted into a Test Utility Method
+<4> After each test case, the Outgoing Emails are cleaned up
+
+As seen in the last callout (4), there is a special way on how to work and verify with the Email Sending functionality.
+
+==== State Verification vs. Behavior Verification
+
+In the Case of the Email Sending, when you think about how it is possible to verify that a particular Email is correctly
+send out to a target Owner this is a little bit of a problematic area to verify. How can the test case assure that an Email was actually send out, because the test case itself is not directly involved in this interaction. Mainly there are the following verification options:
+
+1. check for a received Email in a real existing valid Email Account programmatically
+2. emulate a SMTP server in a test case and verify that a particular message was received from the application
+3. mock the `EmailerAPI` implementation by having a different implementation in the Spring configuration and verify it was called correctly
+
+Those three different approaches to verify can be categorized either as a `State Verification` or a `Behavior Verification`. A behavior verification looks at the two collaborators (can be classes, modules or even systems) and based on the fact that a particular interaction took place derives the conclusion that the verification is correct.
+state verifications on the other hand do not look into how the system interacts with other collaborators, but instead look at the status after the execution happened and checks if the desired state is reached.
+
+In this regard `1.` is a State verification, where `2.` and `3.` are different kinds of behavior verifications.
+
+==== Verification Approach for CUBAs Email Functionality
+
+The `EmailerAPI` has a special method `List<SendingMessage> sendEmailAsync(EmailInfo info);` that is used in the implementation. How in particular this `async` part is handled is, that the API call stores the `EmailInfo` object into a CUBA database Table called `EmailSending` with a particular `SendingStatus`: `QUEUE`. Then an scheduled task is periodically run and will pick up those Emails and tries to send them out. As this happens independently of the thread that executed `emailerAPI.sendEmailAsync(emailInfo)` this process is asynchronous.
+
+This knowledge allows to uncover a forth option of verification:
+
+TIP: 4. Verify database rows to appear in the `EmailSending` CUBA database table that is used by the `EmailerAPI` which represent outgoing Email requests for CUBAs Email functionality.
+
+This option is also a state verification. Generally it is preferable to use State verification as those verifications tend to be more robust and therefore normally increase the maintenance costs of a test case less compared to a behavior verification.
+
+Compared to option `1.` the option `4.` requires much less infrastructure to be in-place. For `1.` the system environment requires:
```

Comment:
If you went for option 3 you would have a chance to demonstrate a bean override in `test-spring.xml`, which I think is a more broad approach to isolate tests from some particular framework implementations...
Maybe at least to explain how to do it, without code examples?

Reply:
> If you went for option 3 you would have a chance to demonstrate a bean override in test-spring.xml

right. I went down the `4.` route, because it is comparably clean to verify and it not mocking. Perhaps this knowledge is indeed worth having. Problem is that I think the guide, as in intro guide is already too deep & intense. 

I was even thinking about cutting scope here before and wanted to ask you for your opinion about that. Putting more examples / concept explanations which is a thing I would like to avoid. But I also don't really have an answer here. It just feels (also compared to the other guides) already quite heavy.

Reply:
If you find it heavy already - it's ok, let's keep it as is.
It's just my idealistic desire to have a complete source of information :)